From joerg.roedel@amd.com  Fri Apr 24 16:58:43 2009
From: Joerg Roedel <joerg.roedel@amd.com>
Date: Fri, 24 Apr 2009 14:57:00 +0200
Subject: driver core: add BUS_NOTIFY_UNBOUND_DRIVER event
To: Greg Kroah-Hartman <gregkh@suse.de>
Cc: Ingo Molnar <mingo@redhat.com>, <iommu@lists.linux-foundation.org>, Joerg Roedel <joerg.roedel@amd.com>
Message-ID: <1240577820-13783-1-git-send-email-joerg.roedel@amd.com>


This patch adds a new bus notifier event which is emitted _after_ a
device is removed from its driver. This event will be used by the
dma-api debug code to check if a driver has released all dma allocations
for that device.

Signed-off-by: Joerg Roedel <joerg.roedel@amd.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/base/dd.c      |    4 ++++
 include/linux/device.h |    2 ++
 2 files changed, 6 insertions(+)

--- a/drivers/base/dd.c
+++ b/drivers/base/dd.c
@@ -320,6 +320,10 @@ static void __device_release_driver(stru
 		devres_release_all(dev);
 		dev->driver = NULL;
 		klist_remove(&dev->p->knode_driver);
+		if (dev->bus)
+			blocking_notifier_call_chain(&dev->bus->p->bus_notifier,
+						     BUS_NOTIFY_UNBOUND_DRIVER,
+						     dev);
 	}
 }
 
--- a/include/linux/device.h
+++ b/include/linux/device.h
@@ -114,6 +114,8 @@ extern int bus_unregister_notifier(struc
 #define BUS_NOTIFY_BOUND_DRIVER		0x00000003 /* driver bound to device */
 #define BUS_NOTIFY_UNBIND_DRIVER	0x00000004 /* driver about to be
 						      unbound */
+#define BUS_NOTIFY_UNBOUND_DRIVER	0x00000005 /* driver is unbound
+						      from the device */
 
 extern struct kset *bus_get_kset(struct bus_type *bus);
 extern struct klist *bus_get_device_klist(struct bus_type *bus);
